File: ble-livedata/src/main/java/no/nordicsemi/android/ble/livedata/ConnectionStateLiveData.java
package no.nordicsemi.android.ble.livedata;

import android.bluetooth.BluetoothDevice;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import no.nordicsemi.android.ble.annotation.DisconnectionReason;
import no.nordicsemi.android.ble.livedata.state.ConnectionState;
import no.nordicsemi.android.ble.observer.ConnectionObserver;

class ConnectionStateLiveData extends LiveData<ConnectionState> implements ConnectionObserver {

	ConnectionStateLiveData() {
		setValue(new ConnectionState.Disconnected(ConnectionObserver.REASON_UNKNOWN));
	}

	@Override
	public void onDeviceConnecting(@NonNull final BluetoothDevice device) {
		setValue(ConnectionState.Connecting.INSTANCE);
	}

	@Override
	public void onDeviceConnected(@NonNull final BluetoothDevice device) {
		setValue(ConnectionState.Initializing.INSTANCE);
	}

	@Override
	public void onDeviceFailedToConnect(
			@NonNull final BluetoothDevice device,
			@DisconnectionReason final int reason) {
		setValue(new ConnectionState.Disconnected(reason));
	}

	@Override
	public void onDeviceReady(@NonNull final BluetoothDevice device) {
		setValue(ConnectionState.Ready.INSTANCE);
	}

	@Override
	public void onDeviceDisconnecting(@NonNull final BluetoothDevice device) {
		setValue(ConnectionState.Disconnecting.INSTANCE);
	}

	@Override
	public void onDeviceDisconnected(
			@NonNull final BluetoothDevice device,
			@DisconnectionReason final int reason) {
		setValue(new ConnectionState.Disconnected(reason));
	}
}

File: ble-livedata/src/main/java/no/nordicsemi/android/ble/livedata/ConnectionStateLiveData.kt
package no.nordicsemi.android.ble.livedata

import android.bluetooth.BluetoothDevice
import androidx.lifecycle.LiveData
import no.nordicsemi.android.ble.livedata.state.ConnectionState
import no.nordicsemi.android.ble.observer.ConnectionObserver

@Suppress("unused")
internal class ConnectionStateLiveData: LiveData<ConnectionState>(
        ConnectionState.Disconnected(reason = ConnectionObserver.REASON_UNKNOWN)
), ConnectionObserver {

    init {
        value = ConnectionState.Disconnected(reason = ConnectionObserver.REASON_UNKNOWN)
    }

    override fun onDeviceConnecting(device: BluetoothDevice) {
        value = ConnectionState.Connecting
    }

    override fun onDeviceConnected(device: BluetoothDevice) {
        value = ConnectionState.Initializing
    }

    override fun onDeviceReady(device: BluetoothDevice) {
        value = ConnectionState.Ready
    }

    override fun onDeviceDisconnecting(device: BluetoothDevice) {
        value = ConnectionState.Disconnecting
    }

    override fun onDeviceDisconnected(device: BluetoothDevice, reason: Int) {
        value = ConnectionState.Disconnected(reason)
    }

    override fun onDeviceFailedToConnect(device: BluetoothDevice, reason: Int) {
        value = ConnectionState.Disconnected(reason)
    }

}